tiatoolbox/models/architecture/sccnn.py

```diff
@@ -0,0 +1,384 @@
+            module_dict["conv1"] = nn.Sequential(
+                nn.Conv2d(
+                    in_channels,
+                    out_channels,
+                    kernel_size=(1, 1),
+                    stride=(1, 1),
+                    padding=0,
+                    bias=True,
+                ),
+                nn.Sigmoid(),
+            )
+            return nn.ModuleDict(module_dict)
+
+        module_dict = OrderedDict()
+        module_dict["l1"] = conv_act_block(num_input_channels, 30, 2)
+        module_dict["pool1"] = nn.MaxPool2d(2, padding=0)
+        module_dict["l2"] = conv_act_block(30, 60, 2)
+        module_dict["pool2"] = nn.MaxPool2d(2, padding=0)
+        module_dict["l3"] = conv_act_block(60, 90, 3)
+        module_dict["l4"] = conv_act_block(90, 1024, 5)
+        module_dict["dropout1"] = nn.Dropout2d(p=0.5)
+        module_dict["l5"] = conv_act_block(1024, 512, 1)
+        module_dict["dropout2"] = nn.Dropout2d(p=0.5)
+        module_dict["sc"] = spatially_constrained_layer1(512, 3)
+
+        self.layer = nn.ModuleDict(module_dict)
+
+    def spatially_constrained_layer2(
+        self, sc1_0: torch.Tensor, sc1_1: torch.Tensor, sc1_2: torch.Tensor
+    ) -> torch.Tensor:
+        """Spatially constrained layer 2.
+
+        Estimates row, column and height for sc2 layer mapping.
+
+        Args:
+            sc1_0 (torch.Tensor):
+                Output of spatially_constrained_layer1 estimating
+                the x position of the nucleus.
+            sc1_1 (torch.Tensor):
+                Output of spatially_constrained_layer1 estimating
+                the y position of the nucleus.
+            sc1_2 (torch.Tensor):
+                Output of spatially_constrained_layer1 estimating
+                the confidence in nucleus detection.
+
+        Returns:
+            :class:`torch.Tensor`:
+                Probability map using the estimates from
+                spatially_constrained_layer1.
+
+        """
+        x = torch.tile(self.xv, dims=[sc1_0.size(0), 1, 1, 1])  # Tile for batch size
+        y = torch.tile(self.yv, dims=[sc1_0.size(0), 1, 1, 1])
+        xvr = (x - sc1_0) ** 2
+        yvc = (y - sc1_1) ** 2
+        out_map = xvr + yvc
+        out_map_threshold = torch.lt(out_map, self.radius).type(torch.float32)
+        denominator = 1 + (out_map / 2)
+        sc2 = sc1_2 / denominator
+        return sc2 * out_map_threshold
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:  # skipcq: PYL-W0221
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+
```

Comment:
same as the suggestion I made for MapDe, I would suggest here to add the preprocessing to the forward method instead of calling it separately before calling the network forward pass. Doing this will make the model more inline with toolbox convention and reduces the risk of user forgetting to put `preproc` in her code.

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/434/commits/dff09b0d1ecbb931f2481bcc1e4d26c8895a38fb

Reply:
I would suggest to add this to `ModelABC`